### PR TITLE
Skip first moment allocation in Adam when beta1=0

### LINKED ISSUE #138248: Adam optimizer without first moment estimate, for less vram
### 🚀 The feature, motivation and pitch

Many GANs like StyleGan use the Adam optimizer with betas=(0.0, 0.999), this means that the first moment estimate is disabled but it still uses a lot of vram for it, is it possible to add an option to stop this from happening and reduce vram usage?

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rajfirke**
Fixes #138248

When beta1=0, the first moment (exp_avg) equals the gradient (m_t = 0*m_{t-1} + 1*g_t = g_t), so allocating a full parameter-sized tensor for it is wasteful. This is common in GAN training where betas=(0, 0.999) is standard (e.g. StyleGAN).

This PR skips exp_avg allocation and uses grad directly in the single-tensor Adam step when beta1=0 and foreach/fused are disabled.

Changes:
1. torch/optim/adam.py: Skip exp_avg allocation in _init_group, use grad directly in _single_tensor_adam
2. test/test_optim.py: Test state savings and correctness vs near-zero beta1

The foreach and fused paths still allocate exp_avg since their tensor-grouping APIs require it. This can be optimized in a follow-up.

cc @janeyx99

### CODE REVIEW

File: torch/optim/adam.py [hunk 1]
```diff
@@ -393,6 +399,8 @@ def _single_tensor_adam(
     else:
         beta1_dict = None
 
+    skip_exp_avg = isinstance(beta1, float) and beta1 == 0.0
```
Comment (Skylion007) [inline torch/optim/adam.py:402]: We should do the same for beta2 for symmetry?

**Comment by rajfirke** [comment]
Good question @Skylion007 ! when beta2=0, exp_avg_sq would also just be grad^2. However, beta2=0 is extremely rare in practice (it disables the adaptive learning rate which is the whole point of Adam). The original issue (#138248) specifically targets beta1=0 which is common in GAN training (e.g. StyleGAN uses betas=(0, 0.999)).

Happy to add beta2=0 support as well if you think it's worth the added complexity, but wanted to keep the initial scope focused on the practical use case. What do you prefer?

**Comment by Skylion007** [comment]
Do we have parity on the CPP API?

**Comment by rajfirke** [comment]
@Skylion007 Re C++ API parityy: the C++ `torch::optim::Adam` in `torch/csrc/api/src/optim/adam.cpp` always allocates `exp_avg` in state initialization regardless of beta1. This PR only optimizes the Python `torch.optim.Adam` eager path (the #138248 use case is Python-only GAN training). Adding the same skip to the C++ API would be a separate change to `AdamParamState` initialization. Happy to file a follow-up if you want parity, or scope this PR to Python-only with a note in the description.

Also pushed a ruff format fix for the failing lintrunner check.
